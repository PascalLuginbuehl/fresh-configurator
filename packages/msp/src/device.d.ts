export interface VoltageMeters {
  id: number;
  voltage: number;
}

export interface AnalogValues {
  voltage: number;
  mahDrawn: number;
  rssi: number;
  amperage: number;
}

export interface RawGpsData {
  fix: number;
  numSat: number;
  lat: number;
  lon: number;
  alt: number;
  speed: number;
  groundCourse: number;
}

export interface BoardInfo {
  boardIdentifier: string;
  boardVersion: number;
  boardType: number;
  targetCapabilities: number;
  targetName: string;
  boardName: string;
  manufacturerId: string;
  signature: number[];
  mcuTypeId: number;
  configurationState: number | undefined;
  sampleRateHz: number | undefined;
}

export type ImuUnit = [number, number, number];

export interface ImuData {
  accelerometer: ImuUnit;
  gyroscope: ImuUnit;
  magnetometer: ImuUnit;
}

export interface Kinematics {
  roll: number;
  pitch: number;
  heading: number;
}

export interface Status {
  cycleTime: number;
  i2cError: number;
  activeSensors: number;
  mode: number;
  profile: number;
}

export interface ExtendedStatus extends Status {
  cpuload: number;
  numProfiles: number;
  rateProfile: number;
  armingDisableCount?: number;
  armingDisableFlags?: number;
}

export interface RcTuning {
  rcRate: number;
  rcExpo: number;
  rollPitchRate: number;
  pitchRate: number;
  rollRate: number;
  yawRate: number;
  dynamicThrottlePid: number;
  throttleMid: number;
  throttleExpo: number;
  dynamicThrottleBreakpoint: number;
  rcYawExpo: number;
  rcYawRate: number;
  rcPitchRate: number;
  rcPitchExpo: number;
  throttleLimitType: number;
  throttleLimitPercent: number;
  rollRateLimit: number;
  pitchRateLimit: number;
  yawRateLimit: number;
}

export interface RcDeadband {
  deadband: number;
  yawDeadband: number;
  altHoldDeadhand: number;
  deadband3dThrottle: number;
}

export enum OSD_VIDEO_TYPES {
  AUTO,
  PAL,
  NTSC
}

export const OSD_VIDEO_VALUE_TO_TYPE = [
  OSD_VIDEO_TYPES.AUTO,
  OSD_VIDEO_TYPES.PAL,
  OSD_VIDEO_TYPES.NTSC
];

export enum OSD_UNIT_TYPES {
  IMPERIAL,
  METRIC
}

export const OSD_UNIT_VALUE_TO_TYPE = [
  OSD_UNIT_TYPES.IMPERIAL,
  OSD_UNIT_TYPES.METRIC
];

export enum OSD_PRECISION_TYPES {
  SECOND,
  HUNDREDTH,
  TENTH
}
export const OSD_PRECISION_VALUE_TO_TYPE = [
  OSD_PRECISION_TYPES.SECOND,
  OSD_PRECISION_TYPES.HUNDREDTH,
  OSD_PRECISION_TYPES.TENTH
];
