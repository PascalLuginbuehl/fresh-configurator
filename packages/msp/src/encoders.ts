import { crc8_dvb_s2_data } from "./utils";

export const encode_message_v1 = (
  code: number,
  data?: number[]
): ArrayBuffer => {
  let bufferOut;
  // always reserve 6 bytes for protocol overhead !
  if (data) {
    var size = data.length + 6,
      checksum = 0;

    bufferOut = new ArrayBuffer(size);
    let bufView = new Uint8Array(bufferOut);

    bufView[0] = 36; // $
    bufView[1] = 77; // M
    bufView[2] = 60; // <
    bufView[3] = data.length;
    bufView[4] = code;

    checksum = bufView[3] ^ bufView[4];

    for (var i = 0; i < data.length; i++) {
      bufView[i + 5] = data[i];

      checksum ^= bufView[i + 5];
    }

    bufView[5 + data.length] = checksum;
  } else {
    bufferOut = new ArrayBuffer(6);
    let bufView = new Uint8Array(bufferOut);

    bufView[0] = 36; // $
    bufView[1] = 77; // M
    bufView[2] = 60; // <
    bufView[3] = 0; // data length
    bufView[4] = code; // code
    bufView[5] = bufView[3] ^ bufView[4]; // checksum
  }
  return bufferOut;
};

export const encode_message_v2 = (
  code: number,
  data?: number[]
): ArrayBuffer => {
  const dataLength = data ? data.length : 0;
  // 9 bytes for protocol overhead
  const bufferSize = dataLength + 9;
  const bufferOut = new ArrayBuffer(bufferSize);
  const bufView = new Uint8Array(bufferOut);
  bufView[0] = 36; // $
  bufView[1] = 88; // X
  bufView[2] = 60; // <
  bufView[3] = 0; // flag
  bufView[4] = code & 0xff;
  bufView[5] = (code >> 8) & 0xff;
  bufView[6] = dataLength & 0xff;
  bufView[7] = (dataLength >> 8) & 0xff;
  for (let ii = 0; ii < dataLength; ii++) {
    bufView[8 + ii] = data![ii];
  }
  bufView[bufferSize - 1] = crc8_dvb_s2_data(bufView, 3, bufferSize - 1);
  return bufferOut;
};
