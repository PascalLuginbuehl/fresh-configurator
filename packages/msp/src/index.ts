export * from "./device";
export {
  ports,
  connections,
  isOpen,
  open,
  close,
  bytesRead,
  bytesWritten,
  packetErrors,
  apiVersion
} from "./serial/connection";
