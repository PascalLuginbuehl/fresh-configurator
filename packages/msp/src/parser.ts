/**
 * The MSP Parser
 *
 * This module is designed to parse bytes and
 * emit data once the full request has been
 * received, parsed, and checked.
 */

/* eslint-disable no-bitwise */
import { Transform } from "stream";
import { MspDataView, crc8DvbS2 } from "./utils";

const SYMBOLS = {
  BEGIN: "$".charCodeAt(0),
  PROTO_V1: "M".charCodeAt(0),
  PROTO_V2: "X".charCodeAt(0),
  FROM_MWC: ">".charCodeAt(0),
  TO_MWC: "<".charCodeAt(0),
  UNSUPPORTED: "!".charCodeAt(0)
};

const enum CONSTANTS {
  PROTOCOL_V1 = 1,
  PROTOCOL_V2 = 2,
  JUMBO_FRAME_MIN_SIZE = 255
}

const enum DECODER_STATES {
  IDLE = 0,
  PROTO_IDENTIFIER = 1,
  DIRECTION_V1 = 2,
  DIRECTION_V2 = 3,
  FLAG_V2 = 4,
  PAYLOAD_LENGTH_V1 = 5,
  PAYLOAD_LENGTH_JUMBO_LOW = 6,
  PAYLOAD_LENGTH_JUMBO_HIGH = 7,
  PAYLOAD_LENGTH_V2_LOW = 8,
  PAYLOAD_LENGTH_V2_HIGH = 9,
  CODE_V1 = 10,
  CODE_JUMBO_V1 = 11,
  CODE_V2_LOW = 12,
  CODE_V2_HIGH = 13,
  PAYLOAD_V1 = 14,
  PAYLOAD_V2 = 15,
  CHECKSUM_V1 = 16,
  CHECKSUM_V2 = 17
}

export interface MspMessage {
  code: number;
  dataView: MspDataView;
  crcError: boolean;
  unsupported: number;
  direction: number;
}

export class MspParser extends Transform {
  private state: DECODER_STATES;
  private messageDirection: number;
  private code: number;
  private message_length_expected: number;
  private message_length_received: number;
  private message_buffer: ArrayBuffer;
  private message_buffer_uint8_view: Uint8Array;
  private message_checksum: number;
  private crcError: boolean;
  private unsupported: number;

  constructor() {
    super();

    this.state = 0;
    this.messageDirection = 1;
    this.code = 0;
    this.message_length_expected = 0;
    this.message_length_received = 0;
    this.message_buffer = new ArrayBuffer(0);
    this.message_buffer_uint8_view = new Uint8Array(0);
    this.message_checksum = 0;
    this.crcError = false;
    this.unsupported = 0;
  }

  // eslint-disable-next-line no-underscore-dangle
  public _transform(
    chunk: Buffer,
    _: string,
    cb: (error?: Error | null, data?: any) => void
  ): void {
    const data = new Uint8Array(chunk);

    for (let i = 0; i < data.length; i += 1) {
      switch (this.state) {
        case DECODER_STATES.IDLE: // sync char 1
          if (data[i] === SYMBOLS.BEGIN) {
            this.state = DECODER_STATES.PROTO_IDENTIFIER;
          }
          break;
        case DECODER_STATES.PROTO_IDENTIFIER: // sync char 2
          switch (data[i]) {
            case SYMBOLS.PROTO_V1:
              this.state = DECODER_STATES.DIRECTION_V1;
              break;
            case SYMBOLS.PROTO_V2:
              this.state = DECODER_STATES.DIRECTION_V2;
              break;
            default:
              console.log(
                `Unknown protocol char ${String.fromCharCode(data[i])}`
              );
              this.state = DECODER_STATES.IDLE;
          }
          break;
        case DECODER_STATES.DIRECTION_V1: // direction (should be >)
        case DECODER_STATES.DIRECTION_V2:
          this.unsupported = 0;
          switch (data[i]) {
            case SYMBOLS.FROM_MWC:
              this.messageDirection = 1;
              break;
            case SYMBOLS.TO_MWC:
              this.messageDirection = 0;
              break;
            case SYMBOLS.UNSUPPORTED:
              this.unsupported = 1;
              break;
            default:
          }
          this.state =
            this.state === DECODER_STATES.DIRECTION_V1
              ? DECODER_STATES.PAYLOAD_LENGTH_V1
              : DECODER_STATES.FLAG_V2;
          break;
        case DECODER_STATES.FLAG_V2:
          // Ignored for now
          this.state = DECODER_STATES.CODE_V2_LOW;
          break;
        case DECODER_STATES.PAYLOAD_LENGTH_V1:
          this.message_length_expected = data[i];

          if (this.message_length_expected === CONSTANTS.JUMBO_FRAME_MIN_SIZE) {
            this.state = DECODER_STATES.CODE_JUMBO_V1;
          } else {
            this.initializeReadBuffer();
            this.state = DECODER_STATES.CODE_V1;
          }

          break;
        case DECODER_STATES.PAYLOAD_LENGTH_V2_LOW:
          this.message_length_expected = data[i];
          this.state = DECODER_STATES.PAYLOAD_LENGTH_V2_HIGH;
          break;
        case DECODER_STATES.PAYLOAD_LENGTH_V2_HIGH:
          this.message_length_expected |= data[i] << 8;
          this.initializeReadBuffer();
          this.state =
            this.message_length_expected > 0
              ? DECODER_STATES.PAYLOAD_V2
              : DECODER_STATES.CHECKSUM_V2;
          break;
        case DECODER_STATES.CODE_V1:
        case DECODER_STATES.CODE_JUMBO_V1:
          this.code = data[i];
          if (this.message_length_expected > 0) {
            // process payload
            if (this.state === DECODER_STATES.CODE_JUMBO_V1) {
              this.state = DECODER_STATES.PAYLOAD_LENGTH_JUMBO_LOW;
            } else {
              this.state = DECODER_STATES.PAYLOAD_V1;
            }
          } else {
            // no payload
            this.state = DECODER_STATES.CHECKSUM_V1;
          }
          break;
        case DECODER_STATES.CODE_V2_LOW:
          this.code = data[i];
          this.state = DECODER_STATES.CODE_V2_HIGH;
          break;
        case DECODER_STATES.CODE_V2_HIGH:
          this.code |= data[i] << 8;
          this.state = DECODER_STATES.PAYLOAD_LENGTH_V2_LOW;
          break;
        case DECODER_STATES.PAYLOAD_LENGTH_JUMBO_LOW:
          this.message_length_expected = data[i];
          this.state = DECODER_STATES.PAYLOAD_LENGTH_JUMBO_HIGH;
          break;
        case DECODER_STATES.PAYLOAD_LENGTH_JUMBO_HIGH:
          this.message_length_expected |= data[i] << 8;
          this.initializeReadBuffer();
          this.state = DECODER_STATES.PAYLOAD_V1;
          break;
        case DECODER_STATES.PAYLOAD_V1:
        case DECODER_STATES.PAYLOAD_V2:
          this.message_buffer_uint8_view[this.message_length_received] =
            data[i];
          this.message_length_received += 1;

          if (this.message_length_received >= this.message_length_expected) {
            this.state =
              this.state === DECODER_STATES.PAYLOAD_V1
                ? DECODER_STATES.CHECKSUM_V1
                : DECODER_STATES.CHECKSUM_V2;
          }
          break;
        case DECODER_STATES.CHECKSUM_V1:
          if (this.message_length_expected >= CONSTANTS.JUMBO_FRAME_MIN_SIZE) {
            this.message_checksum = CONSTANTS.JUMBO_FRAME_MIN_SIZE;
          } else {
            this.message_checksum = this.message_length_expected;
          }
          this.message_checksum ^= this.code;
          if (this.message_length_expected >= CONSTANTS.JUMBO_FRAME_MIN_SIZE) {
            this.message_checksum ^= this.message_length_expected & 0xff;
            this.message_checksum ^=
              (this.message_length_expected & 0xff00) >> 8;
          }
          for (let ii = 0; ii < this.message_length_received; ii += 1) {
            this.message_checksum ^= this.message_buffer_uint8_view[ii];
          }
          this.dispatchMessage(data[i]);
          break;
        case DECODER_STATES.CHECKSUM_V2:
          this.message_checksum = 0;
          this.message_checksum = crc8DvbS2(this.message_checksum, 0); // flag
          this.message_checksum = crc8DvbS2(
            this.message_checksum,
            this.code & 0xff
          );
          this.message_checksum = crc8DvbS2(
            this.message_checksum,
            (this.code & 0xff00) >> 8
          );
          this.message_checksum = crc8DvbS2(
            this.message_checksum,
            this.message_length_expected & 0xff
          );
          this.message_checksum = crc8DvbS2(
            this.message_checksum,
            (this.message_length_expected & 0xff00) >> 8
          );
          for (let ii = 0; ii < this.message_length_received; ii += 1) {
            this.message_checksum = crc8DvbS2(
              this.message_checksum,
              this.message_buffer_uint8_view[ii]
            );
          }
          this.dispatchMessage(data[i]);
          break;
        default:
          console.log(`Unknown state detected: ${this.state}`);
      }
    }
    cb();
  }

  private initializeReadBuffer(): void {
    this.message_buffer = new ArrayBuffer(this.message_length_expected);
    this.message_buffer_uint8_view = new Uint8Array(this.message_buffer);
  }

  private dispatchMessage(expectedChecksum: number): void {
    let dataView: MspDataView;
    if (this.message_checksum === expectedChecksum) {
      // message received, store dataview
      dataView = new MspDataView(
        this.message_buffer,
        0,
        this.message_length_expected
      );
    } else {
      console.log(`code: ${this.code} - crc failed`);
      this.crcError = true;
      dataView = new MspDataView(new ArrayBuffer(0));
    }
    const message: MspMessage = {
      code: this.code,
      dataView,
      crcError: this.crcError,
      unsupported: this.unsupported,
      direction: this.messageDirection
    };

    this.emit("data", message);
    this.reset();
  }

  private reset(): void {
    this.message_length_received = 0;
    this.state = DECODER_STATES.IDLE;
    this.crcError = false;
  }

  // eslint-disable-next-line no-underscore-dangle
  public _flush(cb: () => void): void {
    this.reset();
    cb();
  }
}
