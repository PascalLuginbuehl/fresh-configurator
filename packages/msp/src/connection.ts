/**
 * MSP Serial connections. This allows us to make multiple
 * serial connections, and enables execution of commands
 * on the given connection.
 *
 * The module handles making requests, and dealing with only
 * making one of the same request at a time.
 */

import SerialPort from "@serialport/stream";
import debug from "debug";
import MspDataView from "./dataview";
import { MspMessage, MspParser } from "./parser";
import { encodeMessageV2, encodeMessageV1 } from "./encoders";

import {
  Connection,
  MspCommand,
  ConnectionOptions,
  OpenConnectionFunction,
  OnCloseCallback,
} from "./types";

// Import bindings when we actually need them, so that libraries
// can import this library without the bindings needing to be available
// at import time
const initialiseBindings = async (): Promise<void> => {
  const { default: Binding } = await import("@serialport/bindings");
  SerialPort.Binding = Binding;
};

const log = debug("connection");

const connectionsMap: Record<string, Connection | undefined> = {};

/**
 * Execute the given MspCommand on the given port.
 *
 * Return the data received from the command
 */
export const execute = async (
  port: string,
  { code, data, timeout = 2500 }: MspCommand
): Promise<MspDataView> => {
  const connection = connectionsMap[port];
  if (!connection) {
    throw new Error(`${port} is not open`);
  }

  const sendData = data ? Buffer.from([...data]) : undefined;

  const { parser, serial, requests } = connection;

  const request =
    code > 254
      ? encodeMessageV2(code, sendData)
      : encodeMessageV1(code, sendData);

  const requestKey = request.toString("utf-8");
  let dataRequest = requests[requestKey];

  // Prevent FC lockup by checking if this request is currently being performed
  // and if not, making the request
  if (!dataRequest) {
    dataRequest = new Promise((resolve, reject) => {
      // Throw an error if timeout is reached
      const timeoutId = setTimeout(() => {
        delete requests[requestKey];
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        parser.off("data", onData);
        reject(new Error("Timed out during execution"));
      }, timeout);

      function onData(message: MspMessage): void {
        if (message.code === code) {
          if (message.crcError && connection) {
            connection.packetErrors += 1;
            return;
          }
          log(
            `${request.toJSON().data} response: ${
              Buffer.from(message.data).toJSON().data
            }`
          );

          delete requests[requestKey];
          clearTimeout(timeoutId);

          resolve(message.data);
          parser.off("data", onData);
        }
      }

      parser.on("data", onData);
    });

    requests[requestKey] = dataRequest;

    log(`Writing ${request.toJSON().data} to ${port}`);
    serial.write(request);
    connection.bytesWritten += Buffer.byteLength(request);
  }

  // make every DataView unique to each request, even though
  // they are accessing the same set of data
  return dataRequest.then((responseData) => new MspDataView(responseData));
};

/**
 * Private, used for testing
 */
export const raw = (port: string): SerialPort | undefined =>
  connectionsMap[port]?.serial;

export const reset = (): void => {
  Object.keys(connectionsMap).forEach((port) => {
    delete connectionsMap[port];
  });
};

export const ports = async (): Promise<string[]> => {
  if (!SerialPort.Binding) {
    await initialiseBindings();
  }
  return SerialPort.list().then((data) => data.map(({ path }) => path));
};

/**
 * Close the given port
 */
export const close = async (port: string): Promise<void> => {
  const connection = connectionsMap[port];
  if (!connection) {
    return;
  }
  const { serial } = connection;

  const closePromise = new Promise((resolve) => serial.on("close", resolve));
  serial.close();
  delete connectionsMap[port];

  await closePromise;
};

/**
 * Open a serial connection on the given
 * port, onClose will be emited once the
 * connection is closed
 */
export const open: OpenConnectionFunction = async (
  port: string,
  options?: ConnectionOptions | OnCloseCallback,
  onClose?: OnCloseCallback
) => {
  if (connectionsMap[port]) {
    throw new Error(`${port} is already open`);
  }
  log(`opening ${port}`);

  // Handle overloads
  let onCloseCallback = onClose;
  let connectionOptions = options;

  if (typeof connectionOptions === "function") {
    onCloseCallback = connectionOptions;
    connectionOptions = {};
  }

  if (!SerialPort.Binding) {
    await initialiseBindings();
  }

  const serial = new SerialPort(port, {
    baudRate: 115200,
    ...connectionOptions,
    autoOpen: false,
  });

  await new Promise((resolve, reject) => {
    serial.open((err) => {
      if (err) {
        log(`error opening ${port}`);
        reject(err);
      } else {
        resolve();
      }
    });
  });

  const parser = serial.pipe(new MspParser());
  parser.setMaxListeners(1000000);

  const connection = {
    serial,
    parser,
    requests: {},
    bytesRead: 0,
    bytesWritten: 0,
    packetErrors: 0,
    mspInfo: {
      apiVersion: "0",
      mspProtocolVersion: 0,
    },
  };

  connectionsMap[port] = connection;

  log(`reading API version from ${port}`);
  try {
    const response = await execute(port, { code: 1 });
    connection.mspInfo = {
      mspProtocolVersion: response.readU8(),
      apiVersion: `${response.readU8()}.${response.readU8()}.0`,
    };

    log(`read apiVersion=${connection.mspInfo.apiVersion}`);
  } catch (e) {
    log(`failed to read MSP info from ${port}, failed to open`);
    await close(port);
    throw new Error(`Could not read MSP version from ${port}`);
  }

  // valid connection, setup listeners
  serial.on("data", (data: Buffer) => {
    connection.bytesRead += Buffer.byteLength(data);
  });

  serial.on("error", () => {
    log(`${port} on error received`);
    close(port);
  });

  serial.on("close", () => {
    log(`${port} on close received`);
    close(port);
    onCloseCallback?.();
  });

  connection.bytesRead = 0;
  connection.bytesWritten = 0;
  connection.packetErrors = 0;
};

export const connections = (): string[] => Object.keys(connectionsMap);
export const isOpen = (port: string): boolean => !!connectionsMap[port];

export const bytesRead = (port: string): number =>
  connectionsMap[port]?.bytesRead ?? 0;

export const bytesWritten = (port: string): number =>
  connectionsMap[port]?.bytesWritten ?? 0;

export const packetErrors = (port: string): number =>
  connectionsMap[port]?.packetErrors ?? 0;

export const apiVersion = (port: string): string =>
  connectionsMap[port]?.mspInfo.apiVersion ?? "0.0.0";
