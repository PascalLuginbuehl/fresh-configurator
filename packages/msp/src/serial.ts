/**
 * MSP Serial connections. This allows us to make multiple
 * serial connections, and enables execution of commands
 * on the given connection.
 */
import { MspDataView } from "./utils";
import { MspMessage, MspParser } from "./parser";
import SerialPort = require("serialport");
import { encode_message_v2 } from "./encoders";

interface Connection {
  serial: SerialPort;
  parser: MspParser;
}
const connectionsMap: Record<string, Connection> = {};

export const ports = () =>
  SerialPort.list().then(data => data.map(({ path }) => path));

export const connect = async (
  port: string,
  onClose?: () => void
): Promise<boolean> => {
  if (connectionsMap[port]) {
    return true;
  }

  let serial: SerialPort | undefined;
  await new Promise((resolve, reject) => {
    serial = new SerialPort(port, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });

  if (serial) {
    const parser = serial.pipe(new MspParser());
    parser.setMaxListeners(1000000);

    connectionsMap[port] = {
      serial,
      parser
    };

    serial.on("close", () => {
      delete connectionsMap[port];
      onClose?.();
    });
  }

  return true;
};

export const connections = (): string[] => Object.keys(connectionsMap);
export const isConnected = (port: string) => !!connectionsMap[port];

interface MspCommand {
  code: number;
  data?: number[];
  timeout?: number;
}

/**
 * Execute the given MspCommand on the given port.
 *
 * Return the data received from the command
 */
export const execute = async (
  port: string,
  { code, data, timeout = 5000 }: MspCommand
): Promise<MspDataView> => {
  if (!connectionsMap[port]) {
    throw new Error(`${port} is not open`);
  }

  const { parser, serial } = connectionsMap[port];

  serial.write(Buffer.from(encode_message_v2(code, data)));

  return new Promise((resolve, reject) => {
    const onData = (message: MspMessage) => {
      if (message.code === code) {
        // Copy the data view
        resolve(new MspDataView(message.dataView.buffer()));
        parser.off("data", onData);
      }
    };

    // Throw an error if timeout is reached
    setTimeout(() => {
      parser.off("data", onData);
      reject(new Error("timed out during execution"));
    }, timeout);

    parser.on("data", onData);
  });
};

/**
 * Close the given port
 */
export const close = async (port: string): Promise<void> => {
  connectionsMap[port]?.serial.close();
};
