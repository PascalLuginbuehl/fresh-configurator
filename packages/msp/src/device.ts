import { times } from "rambda";
import semver from "semver";
import codes from "./serial/codes";
import { execute, apiVersion } from "./serial/connection";
import MspDataView from "./serial/dataview";
import {
  VoltageMeters,
  ImuData,
  ImuUnit,
  Kinematics,
  Status,
  ExtendedStatus,
  RcTuning
} from "./device.d";
import { getFeatureBits, Features } from "./features";

export const getVoltages = async (port: string): Promise<VoltageMeters[]> => {
  const data = await execute(port, { code: codes.MSP_VOLTAGE_METERS });
  return times(
    () => ({
      id: data.readU8(),
      voltage: data.readU8() / 10.0
    }),
    3
  );
};

export const getIMUData = async (port: string): Promise<ImuData> => {
  const data = await execute(port, { code: codes.MSP_RAW_IMU });

  const accUnit = (): number => data.read16() / 512;
  const gyroUnit = (): number => data.read16() * (4 / 16.4);
  const mangetUnit = (): number => data.read16() / 1090;

  return {
    accelerometer: times(accUnit, 3) as ImuUnit,
    gyroscope: times(gyroUnit, 3) as ImuUnit,
    magnetometer: times(mangetUnit, 3) as ImuUnit
  };
};

export const getAttitude = async (port: string): Promise<Kinematics> => {
  const data = await execute(port, { code: codes.MSP_ATTITUDE });
  return {
    roll: data.read16() / 10.0, // x
    pitch: data.read16() / 10.0, // y
    heading: data.read16() // z
  };
};

const extractStatus = (data: MspDataView): Status => ({
  cycleTime: data.readU16(),
  i2cError: data.readU16(),
  activeSensors: data.readU16(),
  mode: data.readU32(),
  profile: data.readU8()
});

export const getStatus = async (port: string): Promise<Status> => {
  const data = await execute(port, { code: codes.MSP_STATUS });
  return extractStatus(data);
};

export const getStatusExtended = async (
  port: string
): Promise<ExtendedStatus> => {
  const data = await execute(port, { code: codes.MSP_STATUS_EX });
  const status: ExtendedStatus = {
    ...extractStatus(data),
    cpuload: data.readU16(),
    numProfiles: data.readU8(),
    rateProfile: data.readU8(),
    armingDisableCount: 0,
    armingDisableFlags: 0
  };

  try {
    const byteCount = data.readU8();
    for (let i = 0; i < byteCount; i += 1) {
      data.readU8();
    }

    // Read arming disable flags
    status.armingDisableCount = data.readU8(); // Flag count
    status.armingDisableFlags = data.readU32();
    // eslint-disable-next-line no-empty
  } catch (e) {}
  return status;
};

export const getFeatures = async (
  port: string
): Promise<{ name: Features; enabled: boolean }[]> => {
  const featureBits = getFeatureBits(apiVersion(port));
  const data = await execute(port, { code: codes.MSP_FEATURE_CONFIG });

  const featureMask = data.readU32();
  return Object.entries(featureBits).map(([bit, name]) => ({
    name,
    // eslint-disable-next-line no-bitwise
    enabled: (featureMask >> parseInt(bit, 10)) % 2 !== 0
  }));
};

export const getRcValues = async (port: string): Promise<number[]> => {
  const data = await execute(port, { code: codes.MSP_RC });
  return new Array(data.byteLength / 2).fill(0).map(() => data.readU16());
};

export const getRcTuning = async (port: string): Promise<RcTuning> => {
  const version = apiVersion(port);
  const data = await execute(port, { code: codes.MSP_RC_TUNING });

  const tuning = {
    rcRate: 0,
    rcExpo: 0,
    rollPitchRate: 0,
    pitchRate: 0,
    rollRate: 0,
    yawRate: 0,
    dynamicThrottlePid: 0,
    throttleMid: 0,
    throttleExpo: 0,
    dynamicThrottleBreakpoint: 0,
    rcYawExpo: 0,
    rcYawRate: 0,
    rcPitchRate: 0,
    rcPitchExpo: 0,
    throttleLimitType: 0,
    throttleLimitPercent: 0,
    rollRateLimit: 0,
    pitchRateLimit: 0,
    yawRateLimit: 0
  };

  tuning.rcRate = parseFloat((data.readU8() / 100).toFixed(2));
  tuning.rcExpo = parseFloat((data.readU8() / 100).toFixed(2));

  if (semver.lt(version, "1.7.0")) {
    tuning.rollPitchRate = parseFloat((data.readU8() / 100).toFixed(2));
  } else {
    tuning.rollRate = parseFloat((data.readU8() / 100).toFixed(2));
    tuning.pitchRate = parseFloat((data.readU8() / 100).toFixed(2));
  }

  tuning.yawRate = parseFloat((data.readU8() / 100).toFixed(2));
  tuning.dynamicThrottlePid = parseFloat((data.readU8() / 100).toFixed(2));
  tuning.throttleMid = parseFloat((data.readU8() / 100).toFixed(2));
  tuning.throttleExpo = parseFloat((data.readU8() / 100).toFixed(2));
  tuning.dynamicThrottleBreakpoint = semver.gte(version, "1.7.0")
    ? data.readU16()
    : 0;

  if (semver.gte(version, "1.10.0")) {
    tuning.rcYawExpo = parseFloat((data.readU8() / 100).toFixed(2));
    if (semver.gte(version, "1.16.0")) {
      tuning.rcYawRate = parseFloat((data.readU8() / 100).toFixed(2));
    }
  }

  if (semver.gte(version, "1.37.0")) {
    tuning.rcPitchRate = parseFloat((data.readU8() / 100).toFixed(2));
    tuning.rcPitchExpo = parseFloat((data.readU8() / 100).toFixed(2));
  }

  if (semver.gte(version, "1.41.0")) {
    tuning.throttleLimitType = data.readU8();
    tuning.throttleLimitPercent = data.readU8();
  }

  if (semver.gte(version, "1.42.0")) {
    tuning.rollRateLimit = data.readU16();
    tuning.pitchRateLimit = data.readU16();
    tuning.yawRateLimit = data.readU16();
  }

  return tuning;
};
