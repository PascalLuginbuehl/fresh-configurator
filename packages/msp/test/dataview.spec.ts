import { MspDataView } from "../src";

describe("MspDataView", () => {
  describe("read8", () => {
    it("should read 8 bit integers from the array, in sequence", () => {
      const arrayBuffer = new ArrayBuffer(2);
      const bufferView = new Uint8Array(arrayBuffer);
      bufferView[0] = 15;
      bufferView[1] = 69;

      const dataView = new MspDataView(arrayBuffer, 0, arrayBuffer.byteLength);

      expect(dataView.read8()).toEqual(15);
      expect(dataView.read8()).toEqual(69);
    });

    it("should throw an error if there is no more data to read", () => {
      const arrayBuffer = new ArrayBuffer(2);
      const bufferView = new Uint8Array(arrayBuffer);
      bufferView[0] = 15;

      const dataView = new MspDataView(arrayBuffer, 0, arrayBuffer.byteLength);

      expect(dataView.read8()).toEqual(15);
      expect(dataView.read8()).toEqual(0);
      expect(() => dataView.read8()).toThrowErrorMatchingSnapshot();
    });
  });

  describe("read16", () => {
    it("should read 16 bit integers from the array, in sequence", () => {
      const arrayBuffer = new ArrayBuffer(4);
      const bufferView = new Uint8Array(arrayBuffer);
      bufferView[0] = 15;
      bufferView[1] = 69;
      bufferView[2] = 28;
      bufferView[3] = 96;

      const dataView = new MspDataView(arrayBuffer, 0, arrayBuffer.byteLength);

      expect(dataView.read16()).toEqual(17679);
      expect(dataView.read16()).toEqual(24604);
    });

    it("should throw an error if there is no more data to read", () => {
      const arrayBuffer = new ArrayBuffer(2);
      const bufferView = new Uint8Array(arrayBuffer);
      bufferView[0] = 15;
      bufferView[1] = 69;

      const dataView = new MspDataView(arrayBuffer, 0, arrayBuffer.byteLength);

      expect(dataView.read16()).toEqual(17679);
      expect(() => dataView.read16()).toThrowErrorMatchingSnapshot();
    });
  });
});
