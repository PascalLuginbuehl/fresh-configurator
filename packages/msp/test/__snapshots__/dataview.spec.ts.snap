// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MspDataView read8 should throw an error if there is no more data to read 1`] = `"Data depleated"`;

exports[`MspDataView read16 should throw an error if there is no more data to read 1`] = `"Data depleated"`;
